### PR TITLE
Laboratory work #3, Christina Nigamedzyanova - 21FPL2

### PR DESCRIPTION
**Pull request by christine-ni**

### CODE REVIEW

**Review by marina-kaz — COMMENTED** [review]
Nothing to review just yet

File: lab_3_keywords_textrank/main.py [hunk 1]
```diff
@@ -120,7 +127,10 @@ def _learn_indices(self, tokens: tuple[str, ...]) -> None:
             tokens : tuple[str, ...]
                 sequence of string tokens
         """
-        pass
+        new_tokens = set(list(tokens))
```
[the remark this answers is not in the document]
Comment (christine-ni) [reply lab_3_keywords_textrank/main.py:130]: no, there is no need to do that. I have deleted it already, but forgot to push the changes

File: lab_3_keywords_textrank/main.py [hunk 2]
```diff
@@ -233,7 +266,23 @@ def add_edge(self, vertex1: int, vertex2: int) -> int:
...
+            if i not in self._matrix[0]:
+                self._matrix[0].append(i)
+                self._matrix.append([i])
+        for count in range(len(self._matrix[0])):
+            # вот тут как будто можно переделать
+            for i in range(len(self._matrix[0]) - len(self._matrix[count + 1]) + 1):
+                self._matrix[count + 1].append(0)
+        for i in vertexes:
+            for count, value in enumerate(self._matrix[0]):
+                if i and value in vertexes and i != value:
```
[the remark this answers is not in the document]
Comment (christine-ni) [reply lab_3_keywords_textrank/main.py:282]: yes, I meant "(i and value)" but forgot about parentheses. I corrected the condition now!

File: lab_3_keywords_textrank/main.py [hunk 3]
```diff
@@ -320,7 +393,7 @@ def get_position_weights(self) -> dict[int, float]:
             dict[int, float]
                 position weights for all vertices in the graph
         """
-        pass
+        return self._position_weights
 
 
 class EdgeListGraph:
```
[the remark this answers is not in the document]
Comment (christine-ni) [reply lab_3_keywords_textrank/main.py:395]: done

File: lab_3_keywords_textrank/main.py [hunk 4]
```diff
@@ -233,7 +266,23 @@ def add_edge(self, vertex1: int, vertex2: int) -> int:
...
         """
-        pass
+        if vertex1 == vertex2:
+            return -1
+        vertexes = [vertex1, vertex2]
+        for i in vertexes:
+            if i not in self._matrix[0]:
+                self._matrix[0].append(i)
+                self._matrix.append([i])
+        for count in range(len(self._matrix[0])):
```
[the remark this answers is not in the document]
Comment (marina-kaz) [reply lab_3_keywords_textrank/main.py:275]: please, I would like at least some comments in the code as it is quite perplexing

File: lab_3_keywords_textrank/main.py [hunk 5]
```diff
@@ -278,7 +331,10 @@ def calculate_inout_score(self, vertex: int) -> int:
                 number of incidental vertices
         If vertex is not present in the graph, -1 is returned
         """
-        pass
+        if vertex in self._matrix[0]:
+            v_ind = self._matrix[0].index(vertex)+1
+            return self._matrix[v_ind][1:].count(1)
```
Comment (marina-kaz) [inline lab_3_keywords_textrank/main.py:336]: is there a reason you do not use `sum`?

File: lab_3_keywords_textrank/main.py [hunk 6]
```diff
@@ -385,7 +455,11 @@ def add_edge(self, vertex1: int, vertex2: int) -> int:
...
         """
-        pass
+        if vertex1 == vertex2:
+            return -1
+        self._edges[vertex1] = self._edges.get(vertex1, []) + [vertex2]
+        self._edges[vertex2] = self._edges.get(vertex2, []) + [vertex1]
+        return 0
 
     # Step 7.2
     def is_incidental(self, vertex1: int, vertex2: int) -> int:
```
Comment (marina-kaz) [inline lab_3_keywords_textrank/main.py:465]: in the task description it is specified, that  _if either of vertices is not present in the graph, -1 is returned_

File: lab_3_keywords_textrank/main.py [hunk 7]
```diff
@@ -564,7 +669,7 @@ def get_top_keywords(self, n_keywords: int) -> tuple[int, ...]:
             tuple[int, ...]
                 top n most important tokens in the encoded text
         """
-        pass
+        return [key for (key, value) in sorted(self._scores.items(), key=lambda val: val[1], reverse=True)][:n_keywords]
```
Comment (marina-kaz) [inline lab_3_keywords_textrank/main.py:672]: what if multiple tokens happen to have the same score?

**Review by marina-kaz — COMMENTED** [review]
Great job! I can see substantial progress, keep going!

File: lab_3_keywords_textrank/main.py [hunk 8]
```diff
@@ -51,8 +51,11 @@ def _clean_and_tokenize(self, text: str) -> tuple[str, ...]:
             tuple[str, ...]
                 clean lowercase tokens
         """
-        pass
-
+        for mark in self._punctuation:
+            text = text.replace(mark, '')
+        split_text = text.lower().split()
+        cleaned_text = [i.strip() for i in split_text if i]
```
[the remark this answers is not in the document]
Comment (christine-ni) [reply lab_3_keywords_textrank/main.py:57]: i didn't notice it before, but now i understand that "if i" will always return true, as i use  .split() and all the spaces are deleted

File: lab_3_keywords_textrank/main.py (hunk 5 above)
Comment (christine-ni) [reply lab_3_keywords_textrank/main.py:336]: i just didn't think of using that, i'll consider using sum

File: lab_3_keywords_textrank/main.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (christine-ni) [reply lab_3_keywords_textrank/main.py:275]: it adds zeros to every list in matrix. The length of each list equals the length of self.matrix[0] + 1, and the number of zeros which i need to add is the result of subtraction of the length of self.matrix[0] and the length of the list, which contains a vertex and shows vertexes, incidental to this vertex (i don;t know if the explanation is understandable....)

File: lab_3_keywords_textrank/main.py [hunk 9]
```diff
@@ -564,7 +649,7 @@ def get_top_keywords(self, n_keywords: int) -> tuple[int, ...]:
             tuple[int, ...]
                 top n most important tokens in the encoded text
         """
-        pass
+        return [key for (key, value) in sorted(self._scores.items(), key=lambda val: val[1], reverse=True)]
```
[the remark this answers is not in the document]
Comment (christine-ni) [reply lab_3_keywords_textrank/main.py]: greater numbers should be placed first. I will try to handle this!

File: lab_3_keywords_textrank/main.py (hunk 6 above)
Comment (christine-ni) [reply lab_3_keywords_textrank/main.py:465]: done

File: lab_3_keywords_textrank/main.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (christine-ni) [reply lab_3_keywords_textrank/main.py:57]: done

File: lab_3_keywords_textrank/main.py (hunk 5 above)
Comment (christine-ni) [reply lab_3_keywords_textrank/main.py:336]: done

File: lab_3_keywords_textrank/main.py [hunk 10]
```diff
@@ -233,7 +266,23 @@ def add_edge(self, vertex1: int, vertex2: int) -> int:
...
+        vertexes = [vertex1, vertex2]
+        for i in vertexes:
+            if i not in self._matrix[0]:
+                self._matrix[0].append(i)
+                self._matrix.append([i])
+        for count in range(len(self._matrix[0])):
+            # вот тут как будто можно переделать
+            for i in range(len(self._matrix[0]) - len(self._matrix[count + 1]) + 1):
+                self._matrix[count + 1].append(0)
+        for i in vertexes:
```
[the remark this answers is not in the document]
Comment (christine-ni) [reply lab_3_keywords_textrank/main.py:279]: this cycle replaces zeros with ones. I find the index using the list matrix[0] and check vertexes not to add a loop to the matrix

File: lab_3_keywords_textrank/main.py (hunk 7 above)
Comment (christine-ni) [reply lab_3_keywords_textrank/main.py:672]: greater numbers should be placed first. I will try to handle this!

File: lab_3_keywords_textrank/main.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (marina-kaz) [reply lab_3_keywords_textrank/main.py:275]: i see, thanks!

about the whole function, we've discussed this approach with colleagues and we came to the conclusion that although this approach can be functional it is advised to only use matrix to store 0 or 1, not the actual vertices 

think of a matrix in math: it does not contain names for the rows and columns, it is purely values
please, try to come up with a way to store information about positions-vertex links somehow else 

possibly, i'll write a post about it in the channel... as it is a common issue

File: lab_3_keywords_textrank/main.py (hunk 7 above)
Comment (marina-kaz) [reply lab_3_keywords_textrank/main.py:672]: are you sure? :eyes:

**Review by marina-kaz — COMMENTED** [review]
Keep going and you'll get there!

File: lab_3_keywords_textrank/main.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (christine-ni) [reply lab_3_keywords_textrank/main.py:275]: okay, i'll change this function, thank you for the explanation)

File: lab_3_keywords_textrank/main.py (hunk 7 above)
Comment (christine-ni) [reply lab_3_keywords_textrank/main.py]: i was wrong, lesser numbers should be placed first

File: lab_3_keywords_textrank/main.py [hunk 11]
```diff
@@ -385,7 +455,11 @@ def add_edge(self, vertex1: int, vertex2: int) -> int:
                 0 if edge was added successfully, otherwise -1
         In case of vertex1 being equal to vertex2, -1 is returned as loops are prohibited
         """
-        pass
+        if vertex1 == vertex2:
+            return -1
+        self._edges[vertex1] = self._edges.get(vertex1, []) + [vertex2]
```
Comment (marina-kaz) [inline lab_3_keywords_textrank/main.py:460]: what if these two vertices are already connected?

**Review by marina-kaz — COMMENTED** [review]
Nice!

File: lab_3_keywords_textrank/main.py (hunk 11 above)
Comment (christine-ni) [reply lab_3_keywords_textrank/main.py:460]: the vertices will be added again, and they will be repeated in the value(
i'll try to fix it now!

Comment (marina-kaz) [reply lab_3_keywords_textrank/main.py]: Exactly, and this is not the behaviour we want

**Review by marina-kaz — COMMENTED** [review]
Good job!
A bit of improvement and you'll be all done

File: lab_3_keywords_textrank/main.py (hunk 11 above)
Comment (christine-ni) [reply lab_3_keywords_textrank/main.py]: done

**Pull request closed without merging**